_posts/2016-04-22-responsive-react-components.md

```diff
@@ -0,0 +1,220 @@
+This process was happening over a couple of months with many components being created or updated independent of one another by different developers, it quickly became apparent that between us we had come up with various different solutions at solving the same problem and in the process created some pretty daft technical debt in our shiny new codebase.
+
+So, what was the actual problem we were trying to solve?
+
+The Sky Vegas website needs to render different React components depending on the users current viewport size. A perfect example of this would be the new Navigation.
+
+With lot's of screen real-estate we have the room to render the Navigation component as a full-width horizontal menu:
+![Site Navigation - Large Screen](/images/responsive-react-components_large-screen-nav.png)
+
+However, should the user be navigating on a device with a smaller viewport (such as a mobile phone or tablet), our design team don't have as much screen real-estate to play with and have decided to render the main navigation within a 'burger menu' icon:
+![Site Navigation - Smaller Screen](/images/responsive-react-components_smaller-screen-nav.png)
+
+In short, we needed a way to render different react components depending on the current viewport size of the customer.
+
+A quick search for the problem resolves two seemingly popular options; the classic Javascript `Window.matchMedia()` and the `react-responsive`:
+
+`matchMedia()` Implementation:
+------
+
+``` javascript
+if(window.matchMedia("(min-width: 400px)").matches) {
+    // the viewport is at least 400 pixels wide
+} else {
+    // the viewport is less than 400 pixels wide
+}
+```
+
+`react-responsive` Implementation:
+------
+``` javascript
+var MediaQuery = require('react-responsive');
+
+var A = React.createClass({
+  render: function(){
+    return (
+      <div>
+        <MediaQuery minDeviceWidth={400}>
+            <div>the viewport is at least 400 pixels wide</div>
+        </MediaQuery>
+        <MediaQuery maxDeviceWidth={400}>
+            <div>the viewport is less than 400 pixels wide</div>
+        </MediaQuery>
+      </div>
+    );
+  }
+});
+```
+
+Looking at the implementations above you'll see that each option requires the hard-coding of pixel breakpoints at which you would want to render a different component.
+
+At the late stages of this project and Vegas being a relatvely large squad at the time, this classic code smell was missed and as different engineers cracked on with rendering their own little components into various device-widths both variants of this technique for handling components started to appear in our shiny new codebase.
+
+Upon the realisation of this `foobar`, a tech debt ticket was raised and we set about coming up with a solution to the problem of littering our code with multiple device widths across dozens of components.
```

Comment:
"foobar" in this context should be "fubar" - although replacing it with "After noticing the diversity of approaches" might be better considering what fubar stands for.